nservicebus/sagas/saga-not-found.md

```diff
@@ -29,3 +29,67 @@ The ability to provide an implementation for `IHandleSagaNotFound` is especially
 For example, consider a saga that is used for managing the registration process on the website. After a customer registers, they receive an email with a confirmation link. The system will wait for confirmation for a specific period of time, e.g. 24 hours. If the user doesn't click the link within 24 hours, their data is removed from the system and saga is completed. However, they might decide to click the confirmation link a few days later. In this case, the related saga instance can't be found and an exception will be thrown. By implementing `IHandleSagaNotFound` it is possible to handle the situation differently, e.g. redirect the user to the registration website and ask them to fill out the form again.
 
 The implementation of `IHandleSagaNotFound` should be driven by the business requirements for a specific situation. In some cases the message might be ignored; in others, it might be useful to track whenever that situation happens (e.g. by logging or sending another message). In still other cases, it might make sense to perform a custom compensating action.
+
+## Troubleshooting
+
+Saga not found can occur when:
+
+1. The saga instance does not exist yet
+2. The saga instance already is removed
+
+Sometimes its not always that obvious when saga state will exist or be gone due to race conditions, false assumptions
+
+### Out of order
+
+If the design assumes message A will create the saga and B updates the saga. If message B is received thus processed before A the saga will not yet exist and will be discarded.
+
+Symptoms:
+
+- Saga are created, expected to be completed but this does not always happen.
+
+Mitigation:
+
+- Create the saga for both message A and B.
```

Comment:
```suggestion
- Create the saga for multiple message by using [IAmStartedBy<>(https://docs.particular.net/nservicebus/sagas/#starting-a-saga) with multiple types just as `IHandleMessages<>`. This way out-of-order processing will still create the saga.
  - Do keep in mind that the saga contains logic to "buffer" out of order state until it is in the state to continue.
  - Messages which are a result of a saga handler do not need this. i.e. A reply to a request made by the saga instance does not need to be able to create a new instance.
```
@mauroservienti @mikeminutillo how about this?

Reply:
I think the term "Create the saga" might be confusing. It sounds like the reader needs to do the creation when really we just want them to change the way they declare message handlers. Maybe lead with that?

```suggestion
- Use [IAmStartedBy<>](https://docs.particular.net/nservicebus/sagas/#starting-a-saga) instead of `IHandleMessages<>` for any message type which can originate from outside of the saga. No matter which order messages are delivered in the first one to be processed will create the saga instance.

- Each message handler that can start the saga might need to contain logic to check the saga state and see if it is time to take the next action. 

- Messages which are a result of a saga handler do not need this. i.e. A reply to a request made by saga instance does not need to be able to create a new instance.
```

I wonder if we shouldn't have some examples in this document to make things clearer.

e.g. Imagine a saga that will create schedule shipping after payment has been receieved and stock levels have been checked.

```csharp
class ShippingSchedulePolicy : Saga<ShippingSchedulePolicyData>,
  IAmStartedBy<PaymentReceived>, IAmStartedBy<StockLevelsChecked>
{
  public Task Handle(PaymentReceived message, IMessageHandlerContext context)
  {
    Data.PaymentReceived = true;
    return ScheduleShippingIfReady(context);
  }
  
  public Task Handle(StockLevelsChecked message, IMessageHandlerContext context)
  {
    Data.StockLevelChecked = true;
    return ScheduleShippingIfReady(context);
  }
  
  private async Task ScheduleShippingIfReady(IMessageHandlerContext context)
  {
    if(Data.StockLevelChecked && Data.PaymentReceived)
    {
      await context.Send(new ScheduleShipment { OrderId = Data.OrderId })
        .ConfigureAwait(false);
    }
  }
}
```

Notice that the handler for each message type needs to update the saga state and then check to see if the state is ready before sending the next message. It doesn't matter which order these messages are processed in. When both have been processed, the `ScheduleShipment` message will be sent.

We could do the same for the next bullet point and show a saga that sends a request and handles a response making note of the fact that the response message does not need to be able to start the saga.